pkg/backup/backup.go

```diff
@@ -257,30 +262,147 @@ func (kb *kubernetesBackupper) Backup(log logrus.FieldLogger, backupRequest *Req
+	// store items as they're scraped from the API.
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return errors.Wrap(err, "error creating temp dir for backup")
+	}
+	defer os.RemoveAll(tempDir)
+
+	collector := &itemCollector{
+		log:                   log,
+		backupRequest:         backupRequest,
+		discoveryHelper:       kb.discoveryHelper,
+		dynamicFactory:        kb.dynamicFactory,
+		cohabitatingResources: cohabitatingResources(),
+		dir:                   tempDir,
+	}
+
+	items := collector.getAllItems()
+	log.WithField("progress", "").Infof("Collected %d resources from the Kubernetes API to back up", len(items))
+
+	itemBackupper := &defaultItemBackupper{
+		backupRequest:           backupRequest,
+		tarWriter:               tw,
+		dynamicFactory:          kb.dynamicFactory,
+		discoveryHelper:         kb.discoveryHelper,
+		resticBackupper:         resticBackupper,
+		resticSnapshotTracker:   newPVCSnapshotTracker(),
+		volumeSnapshotterGetter: volumeSnapshotterGetter,
+		itemHookHandler: &defaultItemHookHandler{
+			podCommandExecutor: kb.podCommandExecutor,
+		},
+	}
+	itemBackupper.additionalItemBackupper = itemBackupper
+
+	backedUpGroupResources := map[schema.GroupResource]bool{}
+
+	for i, item := range items {
+		log.WithFields(map[string]interface{}{
+			"progress":  "",
+			"resource":  item.groupResource.String(),
+			"namespace": item.namespace,
+			"name":      item.name,
+		}).Infof("Processing item %d of %d", i+1, len(items))
+
+		// use an anonymous func so we can defer-close/remove the file
+		// as soon as we're done with it
+		func() {
+			var unstructured unstructured.Unstructured
 
-	for _, group := range kb.discoveryHelper.Resources() {
-		if err := gb.backupGroup(group); err != nil {
-			log.WithError(err).WithField("apiGroup", group.String()).Error("Error backing up API group")
+			f, err := os.Open(item.path)
+			if err != nil {
+				log.WithError(errors.WithStack(err)).Error("Error opening file containing item")
+				return
+			}
+			defer f.Close()
+			defer os.Remove(f.Name())
+
+			if err := json.NewDecoder(f).Decode(&unstructured); err != nil {
+				log.WithError(errors.WithStack(err)).Error("Error decoding JSON from file")
+				return
+			}
+
+			if backedUp := kb.backupItem(log, item.groupResource, itemBackupper, &unstructured); backedUp {
+				backedUpGroupResources[item.groupResource] = true
+			}
+		}()
+	}
+
+	// back up CRD for resource if found. We should only need to do this if we've backed up at least
+	// one item for the resource and IncludeClusterResources is nil. If IncludeClusterResources is false
+	// we don't want to back it up, and if it's true it will already be included.
+	if backupRequest.Spec.IncludeClusterResources == nil {
+		for gr := range backedUpGroupResources {
+			kb.backupCRD(log, gr, itemBackupper)
 		}
 	}
 
+	log.WithField("progress", "").Infof("Backed up a total of %d items", len(backupRequest.BackedUpItems))
```

Comment:
TODO fine-tune the language here; the total # of backed-up items may differ from the # of items we "processed" due to:
- additional items being returned from plugins
- exclude-from-backup label applied to a resource
- maybe some other cases too